astropy/io/fits/hdu/image.py

```diff
@@ -246,7 +246,7 @@ def data(self):
 
     @data.setter
     def data(self, data):
-        if 'data' in self.__dict__:
+        if 'data' in self.__dict__ and self.__dict__['data'] is not None:
```

Comment:
It is also set on line 253, which may be why it is set to `True` even if the data was `None`. If the tests are passing it should be good (though I also thought that when the original PR passed :)). Let's leave it as-is for now. Thanks again for catching this!